### DISCUSSION TITLE
Firedrake import error

### DISCUSSION THREAD

**Discussion by edg-13** [firedrakeproject/firedrake#2941] [Firedrake support] [answered]
Hi,

I have installed installed firedrake successfully but I get an error when importing in a jupyter environment. The virtual environment was activated.
![image](https://github.com/firedrakeproject/firedrake/assets/86772327/3885d291-bd19-4ba2-8995-01151ad2f231)

The logs are here:
[make.log](https://github.com/firedrakeproject/firedrake/files/11523232/make.log)
[configure.log](https://github.com/firedrakeproject/firedrake/files/11523233/configure.log)
[firedrake-install.log](https://github.com/firedrakeproject/firedrake/files/11523235/firedrake-install.log)

Would appreciate any help!

**Comment by connorjward** [comment]
It looks like your mpi4py installation is linking to the wrong MPI. Please can you share the output of the command
```
ldd $VIRTUAL_ENV/lib/python*/site-packages/mpi4py/MPI*.so
```

I'm also assuming that you get the same error just with the command `python -c "from firedrake import *"`. This should not be specific to Jupyter notebooks.

@JDBetteridge FYI

**Reply by JDBetteridge** [reply]
I agree, although you should also check that jupyter notebooks is installed inside your virtual-env. If you're using a system or other install of notebooks strange things can happen (like preload a system mpi4py for instance).

**Reply by edg-13** [reply]
Yes, running the import command outside jupyter gives the same behaviour, and jupyter notebooks is installed inside the virtual environment.

This the output of the command
`ldd $VIRTUAL_ENV/lib/python*/site-packages/mpi4py/MPI*.so`

![image](https://github.com/firedrakeproject/firedrake/assets/86772327/0cf71ed0-c3e4-415e-9b9c-a047d8bd5e9d)

**Reply by JDBetteridge** [reply]
Yes, mpi4py has linked against your system mpi `usr/lib/**/libmpi.so`. You can fix this by running `pip uninstall -y mpi4py` followed by:

```
env LDFLAGS="-Wl,-rpath,$VIRTUAL_ENV/src/petsc/default -L$VIRTUAL_ENV/src/petsc/default" MPICC=$VIRTUAL_ENV/src/petsc/default/bin/mpicc MPICXX=$VIRTUAL_ENV/src/petsc/default/bin/mpicxx MPIF90=$VIRTUAL_ENV/src/petsc/default/bin/mpif90 pip install --no-binary mpi4py mpi4py
```

**Reply by edg-13** [reply]
Running both of these commands with the virtual environment activated does not seem to change where the mpi4py has been linked to, or the import error unfortunately.
If you mean running the `pip uninstall -y mpi4py` outside the virtual environment, then I am unable to do so due to lack of permissions, is there a way around this?

**Reply by JDBetteridge** [reply]
I do mean _inside_ the virtual environment. Has `ldd $VIRTUAL_ENV/lib/python*/site-packages/mpi4py/MPI*.so` changed at all?

**Reply by JDBetteridge** [reply]
Oh, I think I missed the `lib` of the end of some of those paths. It should read:
```
env LDFLAGS="-Wl,-rpath,$VIRTUAL_ENV/src/petsc/default/lib -L$VIRTUAL_ENV/src/petsc/default/lib" MPICC=$VIRTUAL_ENV/src/petsc/default/bin/mpicc MPICXX=$VIRTUAL_ENV/src/petsc/default/bin/mpicxx MPIF90=$VIRTUAL_ENV/src/petsc/default/bin/mpif90 pip install --no-binary mpi4py mpi4py
```
I'm on a system with my own PETSc, so can't test this first hand.

**Reply by edg-13** [reply]
Still no luck:
![image](https://github.com/firedrakeproject/firedrake/assets/86772327/3babf774-c03e-45d5-966f-f734e5274c37)

**Reply by JDBetteridge** [reply, marked as answer]
Maybe try with `no-cache`? It might be using a cached wheel stored on your system:
```
env LDFLAGS="-Wl,-rpath,$VIRTUAL_ENV/src/petsc/default/lib -L$VIRTUAL_ENV/src/petsc/default/lib" MPICC=$VIRTUAL_ENV/src/petsc/default/bin/mpicc MPICXX=$VIRTUAL_ENV/src/petsc/default/bin/mpicxx MPIF90=$VIRTUAL_ENV/src/petsc/default/bin/mpif90 pip install --no-cache-dir --no-binary mpi4py mpi4py
```

**Reply by edg-13** [reply]
Yep that was it! Thanks for the help
